README.md
Ext.ux.Help
===========

This extension implements a set of classes used to manage help content
in an Ext JS application. It's designed to be simplistic but easy to use,
and it supports two basic types of help content presentation:  

* tabulated panels with content tree to provide organized and categorized
information
* popup windows for context sensitive help.

All pages are stored as simple text documents that use Markdown tags to
provide basic formatting. Content index is stored in JSON file; at present
there is no automatic index generation so it should be prepared manually.

Tested in:

* MSIE 6+
* Chrome 6+
* Firefox 3.6+
* Opera 11+
* Safari 4+

For usage, see demo application.

This extension is released under GPL 3.0 license.

Commercial use is permitted to the extent that the code/component(s) do NOT
become part of another Open Source or Commercially licensed development library
or toolkit without explicit permission.

Copyright (c) 2011-2012 by Alexander Tokarev, <nohuhu@nohuhu.org>.

This extension depends on Pagedown code, released under MIT license.
See pagedown/LICENSE.txt.

Help icon was taken from Fugue icon set:

(C) 2012 Yusuke Kamiyamane. All rights reserved.
These icons are licensed under a Creative Commons
Attribution 3.0 License.
<http://creativecommons.org/licenses/by/3.0/>

Markdown CSS based on stylesheet by Kevin Burke:
<https://bitbucket.org/kevinburke/markdowncss>


CHANGES
0.03  Aug 14 2012 16:01
- Added support for image shortcuts, analogous to ^ and >

0.02  Aug 10 2012 10:00
- Added an option to align help popup window to HTML element
- Added fade in animation to popup window
- Added setLang method to Panel class

0.01  Aug  5 2012 20:08
- Initial public release


ux/help/Tree.js
/*
 * @private Help index Tree.
 *
 * Copyright (c) 2012 Alexander Tokarev.
 *
 * This component should not be used directly.
 *  
 * This code is licensed under the terms of the Open Source LGPL 3.0 license.
 * Commercial use is permitted to the extent that the code/component(s) do NOT
 * become part of another Open Source or Commercially licensed development library
 * or toolkit without explicit permission.
 * 
 * License details: http://www.gnu.org/licenses/lgpl.html
 */

Ext.define('Ext.ux.help.Tree', {
    extend: 'Ext.tree.Panel',
    
    requires: [
        'Ext.data.TreeStore',
        'Ext.ux.help.Model'
    ],
    
    rootVisible: false,
    
    store: {
        model:    'Ext.ux.help.Model',
        autoLoad: false,
        proxy: {
            type: 'ajax',
            url:  ''
        }
    },
    
    constructor: function(config) {
        var me = this;
        
        Ext.apply(me, config);
        
        Ext.ux.help.Model.imageUrl = me.imageUrl;
        
        me.callParent(arguments);
        
        me.addEvents(
            /**
             * @event load_page
             *
             * Fires when user requests page load by clicking leaf node in the tree.
             *
             * @param {Ext.ux.help.Tree} this Help Tree
             * @param {String} page Page id
             */
            'load_page'
        );
    },
    
    initComponent: function() {
        var me = this;
        
        // Ext.tree.Panel is very rigid; it attempts to load TreeStore
        // regardless of whether we're ready for it or not. So we fool it.
        me.dontSetRoot = true;
        
        me.listeners = Ext.apply(me.listeners || {}, {
            scope:     me,
            itemclick: me.onItemClick
        });
        
        me.callParent();
        
        delete me.dontSetRoot;
    },
    
    setRootNode: function() {
        var me = this;
        
        if ( me.dontSetRoot ) return;
        
        me.callParent();
    },
    
    loadTree: function() {
        var me = this,
            fmt = Ext.String.format;
        
        me.store.setProxy({
            type: 'ajax',
            url:  fmt(me.helpUrl, me.lang) + me.contentIndex
        });
        
        me.store.load();
    },
    
    onItemClick: function(view, node, el, index) {
        var me = this,
            path;
        
        if ( path = node.get('path') ) {
            me.fireEvent('load_page', me, path);
        };
    }
});


ux/help/Model.js
/*
 * @private Help index Model.
 *
 * Copyright (c) 2012 Alexander Tokarev.
 *
 * This component should not be used directly.
 *  
 * This code is licensed under the terms of the Open Source LGPL 3.0 license.
 * Commercial use is permitted to the extent that the code/component(s) do NOT
 * become part of another Open Source or Commercially licensed development library
 * or toolkit without explicit permission.
 * 
 * License details: http://www.gnu.org/licenses/lgpl.html
 */

Ext.define('Ext.ux.help.Model', {
    extend: 'Ext.data.Model',
    
    statics: {
    
        /**
         * @cfg {String} imageUrl Path to directory containing images to be used
         * as icons. Default: '/resources/images/'.
         */
        imageUrl: '/resources/images/'
    },
    
    fields: [{
        name: 'path', type: 'string'
    }, {
        name: 'text', type: 'text'
    }, {
        name: 'children', convert: null
    }, {
        name: 'leaf',
        convert: function(v, rec) {
            return !rec.get('children');
        }
    }, {
        name: 'icon',
        type: 'string',
        convert: function(v, rec) {
            return rec.get('leaf') ? Ext.ux.help.Model.imageUrl + 'book-question.png'
                 :                   undefined
                 ;
        }
    }]
});


ux/help/Pages.js
/*
 * @private Tab panel for help pages.
 *
 * Copyright (c) 2012 Alexander Tokarev.
 *
 * This component should not be used directly.
 *  
 * This code is licensed under the terms of the Open Source LGPL 3.0 license.
 * Commercial use is permitted to the extent that the code/component(s) do NOT
 * become part of another Open Source or Commercially licensed development library
 * or toolkit without explicit permission.
 * 
 * License details: http://www.gnu.org/licenses/lgpl.html
 */

Ext.define('Ext.ux.help.Pages', {
    extend: 'Ext.tab.Panel',
    
    requires: [
        'Ext.Ajax',
        'Ext.window.MessageBox',
        'Ext.ux.help.ContentPanel'
    ],
    
    getPage: function(page) {
        var me = this;
        
        return me.query('panel[pageId="' + page + '"]')[0];
    },
    
    addPage: function(page, text) {
        var me = this,
            panel, baseUrl;
        
        baseUrl = Ext.String.format(me.helpUrl, me.lang).replace(/\/+$/, '/');
        
        panel = new Ext.ux.help.ContentPanel({
            baseUrl:   baseUrl,
            
            pageId:    page,
            text:      text,
            helpTitle: me.helpTitle
        });
        
        me.add(panel);
        me.setActiveTab(panel);
    }
});


ux/help/Panel.js
/*
 * Help Panel, with associated bells and whistles.
 *
 * Version 0.01.
 *  
 * Copyright (c) 2012 Alexander Tokarev.
 *
 * Usage: see demo application.
 *  
 * This code is licensed under the terms of the Open Source LGPL 3.0 license.
 * Commercial use is permitted to the extent that the code/component(s) do NOT
 * become part of another Open Source or Commercially licensed development library
 * or toolkit without explicit permission.
 * 
 * License details: http://www.gnu.org/licenses/lgpl.html
 */

Ext.define('Ext.ux.help.Panel', {
    extend: 'Ext.panel.Panel',
    alias:  'widget.helppanel',
    
    requires: [
        'Ext.layout.container.Border',
        'Ext.ux.help.Manager',
        'Ext.ux.help.Tree',
        'Ext.ux.help.Pages'
    ],
    
    /**
     * @cfg {String} helpUrl Base URL for help components. May include {0}
     * placeholder for language identifier.
     * Default: '/help/{0}/'
     */
    helpUrl: '/help/{0}/',
    
    /**
     * @cfg {String} imageUrl Base URL for images that are used as tree icons.
     * Default: '/resources/images/'
     */
    imageUrl: '/resources/images/',
    
    /**
     * @cfg {String} contentIndex Name of the content index file.
     * Default: index
     */
    contentIndex: 'index',
    
    /**
     * @cfg {String/String[]} defaultPages Page(s) to load by default.
     */
    defaultPages: 'default',
    
    /**
     * @cfg {String} lang (optional) Language identifier for using with localized
     * setups.
     */
    lang: '',
    
    /**
     * @cfg {String} treeTitle Tree panel title. Default: 'Help contents'.
     */
    treeTitle: 'Help contents',
    
    /**
     * @cfg {String} helpTitle Default help panel title, when help page doesn't have
     * h1 header formatted as 'foo bar\n=======\n'. Default: 'Help page'.
     */
    helpTitle: 'Help page',
    
    /**
     * @cfg {String} loadErrorTitle Title for error message box to display when
     * help page can't be loaded.
     * Default: 'Error loading page'
     */
    loadErrorTitle: 'Error loading page',
    
    /**
     * @cfg {String} loadErrorMsg Message text to display when help page can't be loaded.
     * {0} placeholder, if present, will be replaced with status code, and {1} with status
     * text.
     * Default: 'Help page cannot be loaded: {0} {1}.'
     */
    loadErrorMsg: 'Help page cannot be loaded: {0} {1}.',
    
    /**
     * @cfg {Boolean} autoLoadHelp Load help automatically upon startup. This may not be
     * desirable, so you can turn autoloading off and call loadHelp() manually.
     * Default: true.
     */
    autoLoadHelp: true,
    
    /**
     * @cfg {Int} treeWidth Initial content tree panel width. Default: 250 px.
     */
    treeWidth: 250,
    
    layout: 'border',
    
    initComponent: function() {
        var me = this,
            helpUrl, imageUrl;
        
        me.helpUrl  = me.helpUrl.replace(/\/+$/,  '') + '/';
        me.imageUrl = me.imageUrl.replace(/\/+$/, '') + '/';
        
        /**
         * @property {Ext.ux.help.Tree} tree
         */
        me.tree = me.initHelpTree({
            region: 'west',
            width:  me.treeWidth,
            title:  me.treeTitle,
            split:  true,
            
            helpUrl:      me.helpUrl,
            imageUrl:     me.imageUrl,
            contentIndex: me.contentIndex,
            lang:         me.lang
        });
        me.tree.on('load_page', me.onLoadPage, me);

        /**
         * @property {Ext.ux.help.ContentPanel} pages
         */
        me.pages = me.initHelpPages({
            region: 'center',

            helpUrl:        me.helpUrl,
            helpTitle:      me.helpTitle,
            defaultPages:   me.defaultPages,
            lang:           me.lang,
            loadErrorTitle: me.loadErrorTitle,
            loadErrorMsg:   me.loadErrorMsg
        });
        
        me.items = [ me.tree, me.pages ];
        
        me.callParent();
        
        Ext.ux.help.Manager.helpPanel = me;
        Ext.ux.help.Manager.helpTitle = me.helpTitle;
        
        if ( me.autoLoadHelp ) {
            me.loadHelp();
        };
    },
    
    /**
     * @private
     */
    initHelpTree: function(config) {
        return new Ext.ux.help.Tree(config);
    },
    
    /**
     * @private
     */
    initHelpPages: function(config) {
        return new Ext.ux.help.Pages(config);
    },
    
    /**
     * setLang
     *
     * Set help locale to specified identifier.
     */
    setLang: function(lang) {
        var me = this;
        
        me.lang = me.tree.lang = me.pages.lang = lang;
        
        return me;
    },
    
    /**
     * loadHelp
     *
     * Initialize help Panel, optionally replacing default pages with ones from argument.
     *
     * @param {String/String[]} pages (optional) Pages to load at startup
     */
    loadHelp: function(pages) {
        var me = this;
        
        me.tree.loadTree();
        
        Ext.ux.help.Manager.loadPage(pages || me.defaultPages);
    },
    
    /**
     * @private
     */
    onLoadPage: function(tree, page) {
        Ext.ux.help.Manager.loadPage(page);
    }
});


ux/help/Popup.js
/*
 * @private Help popup window.
 *
 * Copyright (c) 2012 Alexander Tokarev.
 *
 * This component should not be used directly.
 *  
 * This code is licensed under the terms of the Open Source LGPL 3.0 license.
 * Commercial use is permitted to the extent that the code/component(s) do NOT
 * become part of another Open Source or Commercially licensed development library
 * or toolkit without explicit permission.
 * 
 * License details: http://www.gnu.org/licenses/lgpl.html
 */

Ext.define('Ext.ux.help.Popup', {
    extend: 'Ext.window.Window',
    alias:  'widget.helppopup',
    
    requires: [
        'Ext.ux.help.ContentPanel'
    ],
    
    /**
     * @cfg {String} page Page id
     */
    
    /**
     * @cfg {String} text Page text
     */
    
    width:  400,
    height: 300,
    
    layout:   'fit',
    closable: true,
    modal:    false,
    
    initComponent: function() {
        var me = this,
            panel;
        
        panel = me.panel = new Ext.ux.help.ContentPanel({
            text:      me.text,
            helpTitle: me.helpTitle
        });
        
        if ( panel.title ) {
            me.title = panel.title;
            delete panel.title;
        };
        
        panel.header = false;
        
        me.items = [ panel ];
        
        me.callParent();
    }
});


ux/help/Manager.js
/*
 * Help content manager. Dispatches page requests, opens popup windows and
 * serves as global point of entry for <a> tag onclick events.
 *
 * Version 0.01.
 *  
 * Copyright (c) 2012 Alexander Tokarev.
 *
 * Usage: see demo application.
 *  
 * This code is licensed under the terms of the Open Source LGPL 3.0 license.
 * Commercial use is permitted to the extent that the code/component(s) do NOT
 * become part of another Open Source or Commercially licensed development library
 * or toolkit without explicit permission.
 * 
 * License details: http://www.gnu.org/licenses/lgpl.html
 */

Ext.define('Ext.ux.help.Manager', {
    singleton: true,
    
    requires: [
        'Ext.window.MessageBox',
        'Ext.ux.help.Popup'
    ],
    
    /**
     * @property {Ext.ux.help.Panel} helpPanel Help panel
     */
    helpPanel: null,
    
    /**
     * @private
     */
    getHelpPanel: function() {
        var me = this;
        
        me.helpPanel = me.helpPanel || Ext.ComponentQuery.query('helppanel')[0];
        
        return me.helpPanel;
    },
    
    /**
     * @private
     */
    noContentPanel: function() {
        Ext.Msg.show({
            title:   'Internal Help error',
            msg:     'Cannot display help page: content panel not found.',
            icon:    Ext.Msg.ERROR,
            buttons: Ext.Msg.OK
        });
        
        return;
    },
    
    /**
     * @private
     */
    getPageUrl: function(page) {
        var me  = this,
            fmt = Ext.String.format,
            panel, url;
        
        panel = me.getHelpPanel();
        
        if ( !panel ) return me.noContentPanel();
        
        url = fmt(panel.helpUrl, panel.lang).replace(/\/+$/, '') + '/' + page;
        
        return url;
    },
    
    /**
     * loadPage
     *
     * Load single page into (optional) tab panel.
     *
     * @param {String/String[]} page Page id(s) to load
     * @param {Ext.tab.Panel} panel (optional) Tab panel to load pages to.
     */
    loadPage: function(page, panel) {
        var me = this,
            pageTab, url;
            
        panel = panel || me.getHelpPanel().pages;
        
        if ( !panel ) return me.noContentPanel();
        
        if ( Ext.isArray(page) ) {
            for ( var i = 0, l = page.length; i < l; i++ ) {
                me.loadPage(page[i], panel);
            };
            
            return;
        };
        
        pageTab = panel.getPage(page);
        
        if ( pageTab ) {
            panel.setActiveTab(pageTab);
            
            return;
        };
        
        url = me.getPageUrl(page);
        
        Ext.Ajax.request({
            url:     url,
            scope:   me,
            success: me.onPageLoadSuccess,
            failure: me.onLoadFailure,
            pageId:  page
        });
    },
    
    /**
     * popup
     *
     * Open a help page in popup window.
     *
     * @param {Object} params
     * @param {String} .page Page id
     * @param {Ext.Element} .el (optional) Element to align popup window to
     * @param {String} .position (optional) Alignment position. Default: 'bl-tl?'.
     * @param {Ext.EventObject} .event (optional) Event that opens this popup
     */
    popup: function(params) {
        var me = this,
            page, url, arg;
    
        if ( !Ext.isObject(params) ) {
            throw 'Ext.ux.help.Manager.popup() requires named arguments';
        };
        
        page = params.page;
        url  = me.getPageUrl(page);
        arg  = Ext.merge({}, params, {
            page:     page,
            position: params.position || 'tl-bl?'
        });
        
        Ext.Ajax.request({
            url:      url,
            scope:    me,
            success:  me.onPopupLoadSuccess,
            failure:  me.onLoadFailure,
            argument: arg
        });
    },
    
    onPageLoadSuccess: function(response, options) {
        var me = this,
            panel;
        
        if ( !response || !Ext.isObject(response) ) {
            me.onPageLoadFailure(response, options);
        };
        
        panel = me.getHelpPanel().pages;
        
        if ( !panel ) return me.noContentPanel();
        
        panel.addPage(options.pageId, response.responseText);
    },
    
    onPopupLoadSuccess: function(response, options) {
        var me = this,
            win, arg, alignEl, alignPos, config;
        
        if ( !response || !Ext.isObject(response) ) {
            me.onPageLoadFailure(response, options);
        };
        
        arg = options.argument;
        
        win = Ext.ComponentQuery.query('window[page="' + arg.page + '"]')[0];
        
        alignEl  = arg.el;
        alignPos = arg.position;
        config   = {
            page:      arg.page,
            text:      response.responseText,
            helpTitle: me.helpTitle
        };
        
        if ( arg.width ) {
            config.width = arg.width;
        };
        
        if ( arg.height ) {
            config.height = arg.height;
        };
        
        if ( win ) {
            win.focus();
        }
        else {
            win = new Ext.ux.help.Popup(config);
            win.showAt(-1000, -1000);
            win.hide();
        };

        if ( alignEl ) {
            win.alignTo(alignEl, alignPos);
        };
        
        win.getEl().fadeIn({
            from:    { opacity: 0 },
            duration: 300
        });
    },
    
    onLoadFailure: function(response, options) {
        var me     = this,
            code   = response.status,
            status = response.statusText,
            panel, msg;
        
        panel = me.getHelpPanel();
        
        if ( !panel ) return me.noContentPanel();
        
        msg = Ext.String.format(panel.loadErrorMsg, code, status);
        
        Ext.Msg.show({
            title:   panel.loadErrorTitle,
            msg:     msg,
            icon:    Ext.Msg.ERROR,
            buttons: Ext.Msg.OK
        });
    }
});


ux/help/ContentPanel.js
/*
 * @private Help content panel.
 *
 * Copyright (c) 2012 Alexander Tokarev.
 *
 * This component should not be used directly.
 *  
 * This code is licensed under the terms of the Open Source LGPL 3.0 license.
 * Commercial use is permitted to the extent that the code/component(s) do NOT
 * become part of another Open Source or Commercially licensed development library
 * or toolkit without explicit permission.
 * 
 * License details: http://www.gnu.org/licenses/lgpl.html
 */

Ext.define('Ext.ux.help.ContentPanel', {
    extend: 'Ext.panel.Panel',
    
    requires: [
        'Ext.util.Format'
    ],
    
    /**
     * @cfg {String} helpTitle Default help page title.
     */
    
    bodyCls:     'ux-help',
    bodyPadding: 8,
    closable:    true,
    autoScroll:  true,
    
    initComponent: function() {
        var me = this,
            title, text, converted;
        
        text = me.text;
        delete me.text;
        
        converted = me.convertText(text);
        
        me.title = converted.title || me.helpTitle;
        me.html  = converted.html;
        
        me.callParent();
    },
    
    convertText: function(text) {
        var me     = this,
            result = {},
            converter, html, match, href;

        converter = new Markdown.Converter();
        
        // Extract page title, if any
        if ( match = /^(.*?)\n[=]+\n/.exec(text) ) {
            result.title = Ext.util.Format.stripTags(match[1]);
            text = text.replace(/^(.*?)\n[=]+\n+/, '');
        };
        
        // Markdown processing mungles all URLs, so we add placeholders
        // at first stage, and then replace them with actual JavaScript later
        text = text.replace(/\[(.*?)\]\(\^([^)]+)\)/g, '[$1](_ux_help_load_page_$2)'  );
        text = text.replace(/\[(.*?)\]\(>([^)]+)\)/g,  '[$1](_ux_help_load_popup_$2)' );
        text = text.replace(/!\[(.*?)\]\(!([^(]+)\)/g, '![$1](_ux_help_insert_img_$2)');
        
        html = converter.makeHtml(text);
        
        html = html.replace(/"_ux_help_load_page_([^"]+)"/g,
                            '"javascript:Ext.ux.help.Manager.loadPage(\'$1\')"'
                           );
        
        html = html.replace(/href="_ux_help_load_popup_([^"]+)"/g,
                            [
                                'href="#" ',
                                'onclick="',
                                    '(function(event) {',
                                        'var ev = new Ext.EventObjectImpl(event),',
                                            'el = new Ext.Element(ev.getTarget());',
                                        
                                        'Ext.ux.help.Manager.popup({',
                                            "page: '$1',",
                                            'el: el,',
                                            'event: ev',
                                        '});',
                                        'return false;',
                                    '})(event)',
                                '"'
                            ].join('')
                           );
        
        html = html.replace(/"_ux_help_insert_img_([^"]+)"/g,
                            '"' + me.baseUrl + '$1"'
                           );
        
        result.html = html;
        
        return result;
    }
});


demo_help.js
/*
 * Help Panel demo application.
 *
 * Version 0.01.
 *  
 * Copyright (c) 2012 Alexander Tokarev.
 *
 * Usage: see below.
 *  
 * This code is licensed under the terms of the Open Source LGPL 3.0 license.
 * Commercial use is permitted to the extent that the code/component(s) do NOT
 * become part of another Open Source or Commercially licensed development library
 * or toolkit without explicit permission.
 * 
 * License details: http://www.gnu.org/licenses/lgpl.html
 */

Ext.Loader.setConfig({
    enabled:        true,
    disableCaching: true,
    paths: {
        'Ext.ux':  'ux'
    }
});

Ext.require([
    'Ext.container.Viewport',
    'Ext.panel.Panel',
    'Ext.layout.container.Border',
    'Ext.ux.help.Panel'
]);

var viewport, panel, tree, content;

Ext.onReady(function() {
    Ext.tip.QuickTipManager.init();

    viewport = new Ext.container.Viewport({
        layout: 'fit',
        
        items: [{
            xtype:        'helppanel',
            helpUrl:      'help',
            imageUrl:     'resources/images',
            contentIndex: 'index',
            defaultPages: [ 'default', 'intro' ],
            autoLoadHelp: true
        }]
    });
});
